### ISSUE TITLE
Cannot use __state when using Factory Method to create Dynamic Prefix Method

### ISSUE THREAD

**Issue by Meigyoku-Thmn** [pardeike/Harmony#184]
**Describe the bug:**
A DynamicMethod object has DeclaringType Property that is always null, this leads to the exception "Object reference not set to an instance of an object." when using __state when using Factory Method to create Dynamic Prefix Method.

**To Reproduce:**
1. The original method and its signature and class:
```csharp
static class Target {
   public static int Sum(int a, int b) {
      return a + b;
   }
}
```
2. The patch code:
```csharp
static class Hook1 {
  static bool Prefix(ref string __state) {
     return false;
  }
}
static class Hook2 {
  static void Sum(int a, int b, ref int __result, ref string __state) {
     __result = a * b;
  }
}
static class Generator {
  static DynamicMethod Prefix(MethodBase method) {
     var original = AccessTools.Method(typeof(Hook1), "Prefix");
     var rs = harmony.Patch(original);
     return rs;
  }
}
class Program {
    static int Test(HarmonyInstance harmony) {
        var original = AccessTools.Method(typeof(Target), "Sum");
        var prefixGen = AccessTools.Method(typeof(Generator), "Prefix");
        var postfix = AccessTools.Method(typeof(Hook2), "Sum");
        harmony.Patch(original, new HarmonyMethod(prefixGen), new HarmonyMethod(postfix));
        Console.WriteLine(Target.Sum(8, 9));
    }
}
```
3. The output of the Harmony debug log: _I don't know where to get this_;
4. The stacktrace or other errors:
```
Object reference not set to an instance of an object.
   at Harmony.MethodPatcher.<>c__DisplayClass8_1.<CreatePatchedMethod>b__2(ParameterInfo patchParam)
   at Harmony.CollectionExtensions.Do[T](IEnumerable`1 sequence, Action`1 action)
   at Harmony.MethodPatcher.<>c__DisplayClass8_0.<CreatePatchedMethod>b__0(MethodInfo prefix)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at Harmony.MethodPatcher.CreatePatchedMethod(MethodBase original, String harmonyInstanceID, List`1 prefixes, List`1 postfixes, List`1 transpilers)
   at Harmony.PatchFunctions.UpdateWrapper(MethodBase original, PatchInfo patchInfo, String instanceID)
   at Harmony.PatchProcessor.Patch()
   at Harmony.HarmonyInstance.Patch(MethodBase original, HarmonyMethod prefix, HarmonyMethod postfix, HarmonyMethod transpiler)
```

**Expected behavior:**
__state and other special parameters work properly.

**Runtime environment:**
 - OS: Windows 7 64 bit
 - .NET version: 4.7.2
 - Harmony version: 1.2.0.1

**Additional context:**
- Where the exception is thrown: https://github.com/pardeike/Harmony/blob/422b16fcd8d86351118aa602fd13b4e171b719e6/Harmony/MethodPatcher.cs#L70

**Comment by pardeike** [comment]
Already fixed in master (upcoming v2.0)

**Comment by Meigyoku-Thmn** [comment]
Ah thanks, I'm looking forward to your v2.0. For now I just change the fix.DeclaringType.FullName into a shared guid for each pair of prefix and postfix as a workaround.
